tests/test_esm_runscripts/test_filedicts.py

```diff
@@ -412,6 +412,32 @@ def test_ln_raises_exception_when_target_path_does_not_exist(simulation_file, fs
 
 # ========== end of ln() tests ==========
 
+def test_check_file_info_is_complete():
+    """Tests for missing variables in the file"""
+    dummy_config = """
+    general:
+        thisrun_dir: "/work/ollie/pgierz/some_exp/run_20010101-20010101"
+        thisrun_work_dir: "/work/ollie/pgierz/some_exp/run_20010101-20010101/work"
+    echam:
+        files:
+            jan_surf:
+                name_in_computer: T63CORE2_jan_surf.nc
+                name_in_work: unit.24
+                path_in_computer: /work/ollie/pool/ECHAM/T63/
+        experiment_input_dir: /work/ollie/pgierz/some_exp/input/echam
+        thisrun_input_dir: /work/ollie/pgierz/some_exp/run_20010101-20010101/input/echam
+    """
+    config = yaml.safe_load(dummy_config)
+
+    # Captures output (i.e. the user-friendly error)
+    with Capturing() as output:
```

Comment:
We don't need `Capturing`. Simply `pytest.raises(SystemExit)` does the job. You can have a look at https://github.com/esm-tools/esm_tools/pull/790

Reply:
That's incorrect Deniz, we do need it. Look at the assert statement below. Miguel is also checking the heading of the user error.

Reply:
I think SystemExit is strong enough. Is there a further need for string parsing?

Reply:
Yes, the new function will report to the user all the issues of a given file at once, so that the user can correct them, and does not need to go syntax error by syntax error. Therefore, checking precisely which error need to be reported it should be part of the test.

Reply:
I still don't get this. I commented out the lines and only left `SystemExit`
```python
# Captures output (i.e. the user-friendly error)
# with Capturing() as output:
with pytest.raises(SystemExit) as error:
    sim_file = esm_runscripts.filedicts.SimulationFile(
        config, "echam.files.jan_surf"
    )

# error needs to occur as the path is not absolute
# assert any(["ERROR: File Dictionaries" in line for line in output])
```
This still gives Green output. Maybe we can talk about that during the meeting.

Reply:
That's because you commented out the test. Comment out the function!

Reply:
`pytest.raises` is the test. Have a look at this: When I correct the path from `pool/ECHAM/T63` to `/pool/ECHAM/T63`, it yields:
```python
        with pytest.raises(SystemExit) as error:
            sim_file = esm_runscripts.filedicts.SimulationFile(
>               config, "echam.files.jan_surf"
            )
E           Failed: DID NOT RAISE <class 'SystemExit'>

/home/ollie/dural/esm_tools/tests/test_esm_runscripts/test_filedicts.py:469: Failed
======================================================================== short test summary info ========
FAILED test_filedicts.py::test_check_path_in_computer_is_abs - Failed: DID NOT RAISE <class 'SystemExit'>
```

But I will bring back the `Capturing` stuff back so that we are all on the same line.

Reply:
Deniz, check now the new unit tests, I think you'll understand now why we need capturing.

Reply:
So, I had thought this was clear last week on Monday when we started, but to definitively write down once how we should be doing this: describe all expected input and all expected output or state changes in the test. Use one single function in between these two to modify the input into the output. Ideally only test one behaviour or state change per test.  

Commenting out part of what the test asserts against (and by the way, at the end of the test) and then wondering why your test does not break doesn't surprise me. You are making the test less comprehensive, so of course it tests less behaviour and is more likely to pass. 

If you wanted to induce a breaking test to discover a bug, you comment out in the function (singular) that modifies the state or expected output, but not in the test. 

If we all as a group decide we need a different broad scale design, then and only then do we redesign and fix the tests. 

But don't let that make you afraid to try stuff out. If you need a new global behaviour and that breaks other peoples stuff, then include in your PR ways to make those previous tests work again. 

With that, enough sprint from me for the day

Reply:
Yes, I check that. One thing I still don't understand is this: unit tests are for the developers and not for the users. If the method is raising an exception and my unit test catches that it is fine. I don't need to parse the error string as an extra effort. The end user will see the error message anyways since it is handled by the `user_error` method. This is why I 
1) load the correct yaml file, 
2) instantiate the `simulation_file` object (no exception is raise)
3) inject buggy line into the `simulation_file` object
4) call the method that is expected to fail / raise exception.
But anyways, this is a difference between the approaches. If it works, it is good.

Reply:
@pgierz, this is why I adopted the test first approach 😃

Reply:
OK then make this into two checks.  Once that it raises an exception. And a second check that ensures the user is correctly informed. 

What Miguel is trying to do here is guard against our most common user complaint. People hate looking at python trackbacks and instead want a beautiful message. The bottom part of this test will ensure that all part of those beautiful messages are included. 

So. Split it. Two tests.

Reply:
And: a system exit with a useful user message is certainly a state change

Reply:
This is why `branch coverage` is important as much as unit testing. Pytest can measure it, as far as I can remember.

Reply:
To adds to this:
> What Miguel is trying to do here is guard against our most common user complaint. People hate looking at python trackbacks and instead want a beautiful message
The function to be tested here tries to get as much information as possible about wrong syntax, and reports all that info at once, so the users don't have to discover one error at a time, one `esm_runscripts --check` at a time. That's what makes this function more sensitive to errors: it needs to assemble a set of user-readable errors and then report the full thing as a user error. I find crucial that the users are reported the correct errors and for that I put together 4 unit tests, testing for each set of errors possible.